ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java

```diff
@@ -179,9 +189,20 @@ public static void process_deposit(BeaconState state, Deposit deposit) {
 
     BLSPublicKey pubkey = deposit.getData().getPubkey();
     UnsignedLong amount = deposit.getData().getAmount();
-    List<BLSPublicKey> validator_pubkeys =
-        state.getValidators().stream().map(Validator::getPubkey).collect(Collectors.toList());
-    if (!validator_pubkeys.contains(pubkey)) {
+
+    SSZList<Validator> validators = state.getValidators();
+    OptionalInt existingIndex;
+    if (pubKeyToIndexMap != null) {
+      Integer cachedIndex = pubKeyToIndexMap.putIfAbsent(pubkey, state.getValidators().size());
+      existingIndex = cachedIndex == null ? OptionalInt.empty() : OptionalInt.of(cachedIndex);
```

Comment:
This took me a bit to get my head around so maybe worth adding a comment.  I believe the logic is:
 * if the validator is already present, just use the cached index
 * if it's not present then we're going to add it to the end of the validator list so it's index will be the current size of that list and we have no `existingIndex`.

That works on the basis that the `pubKeyToIndexMap` contains a key to index mapping for every validator, which holds because this method is private and only called when we're generating the genesis state from a list of ETH1 deposits.

Reply:
Yep, exactly. 
This could be subject to migrate to a global `PubKey -> ValidatorIdex` cache a later optimization stages